File: soc/jh7110/cache.c
#include <stdint.h>
#include <stddef.h>

#define JH7110_L2_CACHE_BASE    0x02010000UL
#define L2_FLUSH64_OFFSET       0x200

/*
 * jh7110_l2_flush_range — flush physical address range to DRAM
 *
 * The SiFive L2 cache controller's Flush64 register accepts a physical
 * address and writes the corresponding dirty cache line (64 bytes) back
 * to DRAM, making it visible to bus masters like the DC8200 display DMA.
 *
 * This is our substitute for Zicbom cbo.flush (not implemented on the U74
 * in JH7110) and Svpbmt non-cacheable page table entries (also absent).
 *
 * Called once after all drawing is complete, before fb_present().
 * For a 1920x1080x4 framebuffer: ~130,000 cache line flushes.
 * At bare-metal MMIO speeds this completes in well under 100ms — fine
 * for a static display. For animation you'd want double-buffering with
 * async flush; for Tutorial-OS a single static frame is sufficient.
 */
void jh7110_l2_flush_range(uintptr_t phys_addr, size_t size)
{
    jh7110_uart_putc('L');   /* confirm entry */
    volatile uint64_t *flush64 =
        (volatile uint64_t *)(0x02010000UL + 0x200);
    uintptr_t line = phys_addr & ~63UL;
    uintptr_t end  = phys_addr + size;
    while (line < end) {
        *flush64 = (uint64_t)line;
        line += 64;
    }
    __asm__ volatile("fence iorw, iorw" ::: "memory");
    jh7110_uart_putc('l');   /* confirm exit */
}

File: soc/jh7110/src/cache.rs
//! Cache management for the StarFive JH7110.
//!
//! Port of `cache.c`.  The SiFive U74 has **no Zicbom** extension, so we
//! cannot use `cbo.flush` / `cbo.clean` / `cbo.inval` instructions.
//!
//! Instead, we use the SiFive L2 cache controller's **Flush64** register
//! at `0x0201_0000 + 0x200`.  Writing a physical address to this register
//! flushes the corresponding 64-byte cache line back to DRAM, making it
//! visible to bus masters like the DC8200 display controller's DMA engine.
//!
//! # When This Is Called
//!
//! Called once after all drawing is complete, before `display_present()`.
//! For a 1920×1080×4 framebuffer: ~130,000 cache line flushes.
//! At bare-metal MMIO speeds this completes in well under 100ms.
//!
//! # Contrast with KyX1
//!
//! The KyX1 (SpacemiT X60) has Zicbom and uses `cbo.clean` loops in
//! `cache.S`.  Same HAL symbol, architecturally different implementation.

use crate::regs;

/// Flush a physical address range from L2 cache to DRAM.
///
/// Uses the SiFive L2 cache controller's Flush64 register.
/// Each write flushes one 64-byte cache line.
pub fn l2_flush_range(phys_addr: usize, size: usize) {
    let flush64 = (regs::L2_CACHE_BASE + regs::L2_FLUSH64_OFFSET) as *mut u64;
    let mut line = phys_addr & !63;
    let end = phys_addr + size;

    while line < end {
        unsafe {
            core::ptr::write_volatile(flush64, line as u64);
        }
        line += 64;
    }

    // Fence to ensure all flushes complete before we continue
    crate::cpu::dsb();
}

/// Clean the data cache for a given range (HAL-compatible name).
///
/// Maps to [`l2_flush_range`] on JH7110.
pub fn clean_dcache_range(start: usize, size: usize) {
    l2_flush_range(start, size);
}
